Add bot detection to skip event recording in analytics tracking

chihsuan opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->


## Proposed changes:
<!--- Explain what functional changes your PR includes -->

When I looked at [WOOA7S-568](https://linear.app/a8c/issue/WOOA7S-568/woocommerceanalytics-product-purchase-tracks-event-is-getting-rejected), I've noticed that a lot of bot events. This PR adds bot detection to skip event recording in analytics tracking to prevent unnecessary data collection from bots.

<img width="533" height="254" alt="Screenshot 2025-10-20 at 15 42 21" src="https://github.com/user-attachments/assets/1be860ec-698c-4012-8d87-eddf114c6456" />


### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Use the following code snippet to log the events:

```php
 function log_remote_pixels( $preempt, $parsed_args, $url ) {
	if ( strpos( $url, WC_Tracks_Client::PIXEL ) === 0 ||  strpos( $url, 'https://pixel.wp.com/w.gif' ) === 0  ) {
		$parsed_url = wp_parse_url( $url );
		parse_str( $parsed_url['query'], $params );
		error_log( $url );
		error_log( $params['_en'] . ' -- ' . print_r( $params, true ));
	}

	return $preempt;
}

add_action( 'pre_http_request', 'log_remote_pixels', 10, 3 );
```

* Go to your Shop page 
* Add a product to the cart
* Confirm that `add_to_cart` event is logged
* Change user agent to `Mozilla/5.0 AppleWebKit/537.36 (KHTML, like Gecko; compatible; GPTBot/1.2; +https://openai.com/gptbot)` (See https://developer.chrome.com/docs/devtools/device-mode/override-user-agent)
* Add a product to the cart again
* Confirm that `add_to_cart` event is not logged

chihsuan committed:
Add bot detection to skip event recording in analytics tracking

- Implemented a method to identify bot requests based on user agent.
- Updated event recording logic to skip events from detected bots.
- Fixed a typo in the method documentation for event name parameter.

chihsuan committed:
changelog

chihsuan committed:
Update bot detection regex in analytics tracking to improve accuracy

- Replaced the existing user agent regex with a more comprehensive pattern to identify a wider range of bots.
- Adjusted regex delimiters to prevent issues with unescaped characters.

kangzj reviewed (commented):
I don't know whether it makes sense to update and use the existing bot detection code, which Jetpack and Jetpack Stats both are using, including in log parsing processes.

https://github.com/Automattic/jetpack/blob/79b082cef4ed1694ac088d0d6fc31bb526486dcd/projects/packages/device-detection/src/class-user-agent-info.php#L1663

However the changes looks good to me!

chihsuan commented:
> I don't know whether it makes sense to update and use the existing bot detection code, which Jetpack and Jetpack Stats both are using, including in log parsing processes.
> 
> https://github.com/Automattic/jetpack/blob/79b082cef4ed1694ac088d0d6fc31bb526486dcd/projects/packages/device-detection/src/class-user-agent-info.php#L1663
> 
> However the changes looks good to me!

I wasn’t aware we already had that package. Thanks for pointing it out!  @kangzj I agree, it makes sense to update and use the existing bot detection code.

chihsuan committed:
Add new bot user agents to detection list

- Expanded the list of recognized bot user agents in the User_Agent_Info class to include various new bots, enhancing the accuracy of bot detection.
- This update includes well-known bots such as gptbot, claudebot, and others, improving overall functionality.

chihsuan committed:
Integrate Device Detection for Bot Tracking

- Added the Device Detection package to enhance bot detection capabilities in analytics tracking.
- Updated the logic to utilize the existing bot detection method, improving accuracy in skipping event recording for bot requests.
- Revised changelog entry for clarity on the changes made.

chihsuan committed:
Update changelog

kangzj commented:
I just had a call with @gibrown where we talked about how to deal with bot traffic. In general, we thought we should probably record the traffic, but filtering/processing it from the server side as we need the visibility of bots for various reasons. Hi @gibrown do you have comments here? Thanks

chihsuan commented:
> I just had a call with @gibrown this morning where we talked about how to deal with bot traffic. In general, we thought we should probably record the traffic, but filtering/processing it from the server side as we need the visibility of bots for various reasons. Hi @gibrown do you have comments here? Thanks

That approach works for me! Let me know what we want to do for bot traffic, and I can update the code accordingly. 🙂 

I also wanted to point out that I added this logic because we already block bot traffic in `w.js` and `s.js`. Therefore, if we're planning to track traffic, we should ensure our approach remains unified and consistent.

chihsuan committed:
Merge branch 'trunk' into update/stop-record-events-bot

chihsuan committed:
Update composer.lock

chihsuan committed:
changelog
